openedx/core/djangoapps/content/search/api.py

```diff
@@ -327,11 +327,20 @@ def rebuild_index(status_cb: Callable[[str], None] | None = None) -> None:
         ])
         # Mark which attributes are used for keyword search, in order of importance:
         client.index(temp_index_name).update_searchable_attributes([
+            # Keyword search does _not_ search the course name, course ID, breadcrumbs, block type, or other fields.
             Fields.display_name,
             Fields.block_id,
             Fields.content,
             Fields.tags,
-            # Keyword search does _not_ search the course name, course ID, breadcrumbs, block type, or other fields.
+            # If we don't list the following sub-fields _explicitly_, they're only sometimes searchable - that is, they
+            # are searchable only if at least one document in the index has a value. If we didn't list them here and,
+            # say, there were no tags.level3 tags in the index, the client would get an error if trying to search for
+            # these sub-fields: "Attribute `tags.level3` is not searchable."
+            Fields.tags + "." + Fields.tags_taxonomy,
+            Fields.tags + "." + Fields.tags_level0,
+            Fields.tags + "." + Fields.tags_level1,
+            Fields.tags + "." + Fields.tags_level2,
+            Fields.tags + "." + Fields.tags_level3,
```

Comment:
Hmm.. I thought we only supported a [3-level tag hierarchy](https://github.com/openedx/openedx-learning/blob/ca4d428f8698ab17edf991adb19bb316be443714/openedx_tagging/core/tagging/models/base.py#L28), but this PR uses 4 levels, and I can see 4 levels in ESDC:

![image](https://github.com/openedx/edx-platform/assets/7556571/b5757fa4-7403-46f4-a498-10590b71e200)

Do we need to update openedx-learning too?

Reply:
Yeah, I've been a bit inconsistent about that. We only officially support 3 levels, though at the core the code will support a higher number of levels - it's just that some search and listing features only work on three (or in this case 4) levels. It would be a good change in the future to drop this fourth level from the search engine and to prevent the import or creation of taxonomies with 4 levels (or support 4 levels). For now I don't think it's a bug we need to fix though.